File: backend/src/lyrics/worker.client.ts
import { Injectable } from '@nestjs/common';
import { NatsService } from '../bus/nats.service.js';
import { SUBJECTS } from '../bus/subjects.js';

export interface RankCandidate {
  idx: number;
  source: string;
  snippet: string;
}

export interface RankResult {
  best_idx: number;
  score: number;
  scores?: Array<{ idx: number; score: number }>;
}

export interface TranscribeResult {
  syncedLrc: string | null;
  plainText: string | null;
  language: string;
}

@Injectable()
export class WorkerClient {
  constructor(private readonly nats: NatsService) {}

  detectLanguage(text: string): Promise<{ language: string; confidence: number } | null> {
    return this.nats.request(SUBJECTS.aiDetectLanguage, { text }, 15_000);
  }

  async generateSearchQueries(artist: string, title: string): Promise<string[]> {
    const res = await this.nats.request<{ queries: string[] }>(
      SUBJECTS.aiSearchQueries,
      { artist, title },
      40_000,
    );
    const queries = res?.queries?.filter((q) => q?.trim()) ?? [];
    return queries.length ? queries : [`${artist} ${title}`.trim()];
  }

  rankLyrics(
    artist: string,
    title: string,
    candidates: RankCandidate[],
  ): Promise<RankResult | null> {
    if (!candidates.length) return Promise.resolve(null);
    return this.nats.request(SUBJECTS.aiRankLyrics, { artist, title, candidates }, 60_000);
  }

  transcribeAudio(
    audioUrl: string,
    language?: string,
    initialPrompt?: string,
  ): Promise<TranscribeResult | null> {
    return this.nats.request(
      SUBJECTS.aiTranscribe,
      { audio_url: audioUrl, language, initial_prompt: initialPrompt },
      180_000,
      { throwOnError: true },
    );
  }

  async encodeTextMulan(text: string): Promise<number[] | null> {
    const res = await this.nats.request<{ vector: number[] }>(
      SUBJECTS.aiEncodeTextMulan,
      { text },
      15_000,
    );
    return res?.vector ?? null;
  }
}

File: backend/src/modules/lyrics/worker_client.rs
use std::sync::Arc;
use std::time::Duration;

use serde::{Deserialize, Serialize};

use crate::bus::nats::NatsService;
use crate::bus::subjects::subjects;
use crate::error::AppResult;

#[derive(Debug, Clone, Serialize)]
pub struct RankCandidate {
    pub idx: usize,
    pub source: String,
    pub snippet: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct RankResult {
    pub best_idx: usize,
    pub score: f32,
}

#[derive(Debug, Clone, Deserialize)]
pub struct TranscribeResult {
    #[serde(rename = "syncedLrc")]
    pub synced_lrc: Option<String>,
    #[serde(rename = "plainText")]
    pub plain_text: Option<String>,
    pub language: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct LangResult {
    pub language: String,
    pub confidence: f32,
}

pub struct WorkerClient {
    nats: Arc<NatsService>,
}

impl WorkerClient {
    pub fn new(nats: Arc<NatsService>) -> Arc<Self> {
        Arc::new(Self { nats })
    }

    pub async fn detect_language(&self, text: &str) -> AppResult<Option<LangResult>> {
        self.nats
            .request::<_, LangResult>(
                subjects::AI_DETECT_LANGUAGE,
                &serde_json::json!({ "text": text }),
                Duration::from_secs(15),
                false,
            )
            .await
    }

    pub async fn generate_search_queries(
        &self,
        artist: &str,
        title: &str,
    ) -> AppResult<Vec<String>> {
        #[derive(Deserialize)]
        struct Resp {
            queries: Option<Vec<String>>,
        }
        let res: Option<Resp> = self
            .nats
            .request(
                subjects::AI_SEARCH_QUERIES,
                &serde_json::json!({ "artist": artist, "title": title }),
                Duration::from_secs(40),
                false,
            )
            .await?;
        let queries: Vec<String> = res
            .and_then(|r| r.queries)
            .map(|v| v.into_iter().filter(|q| !q.trim().is_empty()).collect())
            .unwrap_or_default();
        if queries.is_empty() {
            let fallback = format!("{artist} {title}").trim().to_string();
            if fallback.is_empty() {
                Ok(Vec::new())
            } else {
                Ok(vec![fallback])
            }
        } else {
            Ok(queries)
        }
    }

    pub async fn rank_lyrics(
        &self,
        artist: &str,
        title: &str,
        candidates: &[RankCandidate],
    ) -> AppResult<Option<RankResult>> {
        if candidates.is_empty() {
            return Ok(None);
        }
        self.nats
            .request(
                subjects::AI_RANK_LYRICS,
                &serde_json::json!({
                    "artist": artist,
                    "title": title,
                    "candidates": candidates,
                }),
                Duration::from_secs(60),
                false,
            )
            .await
    }

    pub async fn transcribe_audio(
        &self,
        audio_url: &str,
        language: Option<&str>,
        initial_prompt: Option<&str>,
    ) -> AppResult<Option<TranscribeResult>> {
        self.nats
            .request(
                subjects::AI_TRANSCRIBE,
                &serde_json::json!({
                    "audio_url": audio_url,
                    "language": language,
                    "initial_prompt": initial_prompt,
                }),
                Duration::from_secs(180),
                true,
            )
            .await
    }

    pub async fn encode_text_mulan(&self, text: &str) -> AppResult<Option<Vec<f32>>> {
        #[derive(Deserialize)]
        struct Resp {
            vector: Option<Vec<f32>>,
        }
        let res: Option<Resp> = self
            .nats
            .request(
                subjects::AI_ENCODE_TEXT_MULAN,
                &serde_json::json!({ "text": text }),
                Duration::from_secs(15),
                false,
            )
            .await?;
        Ok(res.and_then(|r| r.vector))
    }
}
